### PR TITLE
Adds possibility to display file name on file upload

### PR DESCRIPTION
**Pull request by tomas862**
# About

Also on multiple file names it displays number of files and text retrieved from data attribute.

input1

numeric_files

Related with issue https://github.com/PrestaShop/PrestaShop/issues/12360

## How to test

Clone the repository, open the root index.html. Look for this part:

The left "Choose file" input is a regular one, which displays the name of the file being uploaded.
The right "Choose files" input allows multiple files uploads by doing this: holding shift + mouse click or ctrl + mouse click. If you do this, it will display the number of files.

### CODE REVIEW

File: js/prestashop-ui-kit.js [hunk 1]
```diff
@@ -118,6 +118,27 @@ $.fn.pstooltip = $.fn.tooltip;
...
+
+      if (1 === filesCount) {
+        $label.text(files[0].name);
+      } else if (1 < filesCount) {
+        var multipleFilesTranslationText = $input.attr('data-multiple-files-text');
+
+        if (typeof multipleFilesTranslationText !== 'undefined') {
+          $label.text(filesCount + ' ' + multipleFilesTranslationText);
+        } else {
+          $label.text(filesCount + ' files');
```
Comment (eternoendless) [inline js/prestashop-ui-kit.js]: Consider extracting the logic that decides what text you'll put into your label, so that you end up having:

```js
$label.text(getLabelText(files))
```

File: js/prestashop-ui-kit.js [hunk 2]
```diff
@@ -118,6 +118,27 @@ $.fn.pstooltip = $.fn.tooltip;
       });
     });
   }
+  
+  var initInputFileChangeEvent = function () {
+    jQuery('.custom-file input').on('change', function () {
```
Comment (eternoendless) [inline js/prestashop-ui-kit.js]:
```suggestion
    jQuery('.custom-file-input').on('change', function () {
```

File: js/prestashop-ui-kit.js [hunk 3]
```diff
@@ -118,6 +118,27 @@ $.fn.pstooltip = $.fn.tooltip;
       });
     });
   }
+  
+  var initInputFileChangeEvent = function () {
```
Comment (eternoendless) [inline js/prestashop-ui-kit.js]: Consider putting this in a separate module.

File: index.html [hunk 4]
```diff
@@ -2117,7 +2117,7 @@ <h2>Forms</h2>
                         <br><br>
 
                         <div class="custom-file">
-                            <input type="file" class="custom-file-input" id="file3">
+                            <input type="file" class="custom-file-input" id="file3" multiple data-multiple-files-text="files">
```
Comment (eternoendless) [inline index.html]: > `data-multiple-files-text="files"`

This is going to be a royal pain in the ass to localize.

First, we need to pass a transchoice compatible string here (see [Pluralization](https://symfony.com/doc/3.4/translation.html#pluralization))

```
{0} no files|{1} 1 file|]1,Inf[ %count% files
```

Then, it should be parsed according to the number of elements using something like this:
https://github.com/rmariuzzo/Lang.js/blob/master/src/lang.js#L178-L228

Comment (tomas862) [reply index.html]: Hi @eternoendless  - just want to confirm the task and check if I understand everything correctly.

1. I'll pass a translation withing the data attribute like:
```twig
<input data-pluralised-text="{{ '{0} no files|{1} 1 file|]1,Inf[ %count% files'|trans({}, 'Admin.Global') }}">
```

2. I also need a locale so I'll pass it with another data attribute
3. I'll use the code from the link you have provided - like 90% of the stuff is similar

Comment (eternoendless) [reply index.html]: > 2. I also need a locale so I'll pass it with another data attribute

No, you don't need a locale because by using "trans" in Twig the whole string is translated without making a choice on what pattern to use.

For example

```twig
{{ '{0} no files|{1} 1 file|]1,Inf[ %count% files'|trans({}, 'Admin.Global') }}"
```

Should output in french:

```
{0} aucun fichier|{1} 1 fichier|]1,Inf[ %count% fichiers
```

Then we'll use that with transchoice to figure out which of the three choices we'll use, according to the number of `files`.

eg. with "1" file:

```
1 fichier
```

UI kit of course knows nothing of translation, it should only be able to make a choice out of the provided, **already translated** pattern.

Comment (tomas862) [reply index.html]: @eternoendless  I have analysed a bit this part https://github.com/rmariuzzo/Lang.js/blob/master/src/lang.js#L494 and tried to imagine how it would look like in my own language to translate it:

there are two plural forms in Lithuanian (LT) and one singular form:

for numers like 21,31,41,51,61 etc... it uses singular form **failas**
for numbers like 22,23,24,25... it uses **failai**
for numbers like 30,40,50,60,70... it uses **failų**

its impossible to define all rules  so here its the certain math applied , but only for my language case. The code in https://github.com/rmariuzzo/Lang.js/blob/master/src/lang.js#L494 solves it but it requires to know the iso code.

My question is - are we really don't need to know the locale because without using the locale we won't get the correct results in all possible cases. :thinking:

Comment (eternoendless) [reply index.html]: I guess we'll have to support the locale code too. Just make it optional and fall back to 'en' if not provided.

**Comment by tomas862** [comment]
@eternoendless done the requested changes

With the translatable file message I tested manually some edge cases but maybe I need to add some kind of tests to it, wdyt?

File: js/utils/Translator.js [hunk 5]
```diff
@@ -0,0 +1,393 @@
+(function (root, factory) {
```
Comment (eternoendless) [inline js/utils/Translator.js]: We should add a copyright notice at the top where we keep attribution to the original authors.

File: js/utils/Translator.js [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+  if (typeof define === 'function' && define.amd) {
+    define('Translator', factory);
```
Comment (eternoendless) [inline js/utils/Translator.js]: Not sure about the name "translator" because there's not actually any translation involved. I think "Pluralizer" would be more appropriate.

File: js/utils/Translator.js [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+  else if (typeof module === 'object' && module.exports) {
+    // Node. Does not work with strict CommonJS, but
+    // only CommonJS-like environments that support module.exports,
+    // like Node.
+    module.exports = factory();
+  }
+  else {
+    root.Translator = factory();
+  }
+}(this, function () {
```
Comment (eternoendless) [inline js/utils/Translator.js]: I don't think we need the block above since we use Webpack for bundling. You can simply add this at the bottom:

```js
export default Pluralizer;
```

and then import it on the other script:

```js
import Pluralizer from 'utils/pluralizer.js'
```

File: js/events/file-input.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    var files = $input[0].files;
+    var $label = $input.next('label');
+
+    $label.text(
+      getTextByFilesCount(files, $input.attr('data-multiple-files-text'), $input.attr('data-locale'))
+    );
+  });
+};
+
+module.exports = initInputFileChangeEvent;
```
Comment (eternoendless) [inline js/events/file-input.js]: use `export default` instead.

File: js/events/file-input.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
+var Translator = require('../utils/Translator.js');
```
Comment (eternoendless) [inline js/events/file-input.js]: I think it's better to use minuscules for js file names.

File: js/prestashop-ui-kit.js [hunk 10]
```diff
@@ -3,6 +3,7 @@ import 'bootstrap';
 import 'select2';
 import 'jquery.growl';
 import 'pstagger';
+var initInputFileChangeEvent = require('./events/file-input.js');
```
Comment (eternoendless) [inline js/prestashop-ui-kit.js]: Use `import` instead or require.

Also, maybe it would be better to have `/components/file-input` instead of `/events/file-input`.

File: js/prestashop-ui-kit.js [hunk 11]
```diff
@@ -165,6 +166,7 @@ $.fn.pstooltip = $.fn.tooltip;
         initAlerts();
         initToolTips();
         initSearchBar();
+        initInputFileChangeEvent();
```
Comment (eternoendless) [inline js/prestashop-ui-kit.js]: Maybe this should be `initInputFile()`?

**Comment by eternoendless** [comment]
> With the translatable file message I tested manually some edge cases but maybe I need to add some kind of tests to it, wdyt?

I think it's a good idea but we can do that later. I think the component will have to be externalized into a project of its own, including unit tests.

**Comment by tomas862** [comment]
@eternoendless done requested changes - thank you!

File: js/utils/pluralizer.js [hunk 12]
```diff
@@ -0,0 +1,386 @@
+/*!
+ *  Lang.js for Laravel localization in JavaScript.
+ *
+ *  @version 1.1.12
+ *  @license MIT https://github.com/rmariuzzo/Lang.js/blob/master/LICENSE
+ *  @site    https://github.com/rmariuzzo/Lang.js
+ *  @author  Rubens Mariuzzo <rubens@mariuzzo.com>
+ */
```
Comment (eternoendless) [inline js/utils/pluralizer.js]: I'd change this block as follows

```js
/*!
 * This file has been derived from "Lang.js - Laravel Translator class in JavaScript", v1.1.12 by Rubens Mariuzzo <rubens@mariuzzo.com>, distributed under the MIT licence.
 *
 * @site https://github.com/rmariuzzo/Lang.js
 */
```

File: package-lock.json [hunk 13]
```diff
@@ -1,8718 +0,0 @@
-{
-  "name": "prestakit",
```
Comment (eternoendless) [inline package-lock.json]: This file must not be deleted.

File: js/components/file-input.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+ * just by drag and drop action applied.
+ */
+var initInputFile = function () {
+  jQuery('.custom-file-input').on('change', function () {
+    var $input = jQuery(this);
+    var files = $input[0].files;
+    var $label = $input.next('label');
+
+    $label.text(
+      getTextByFilesCount(files, $input.attr('data-multiple-files-text'), $input.attr('data-locale'))
```
Comment (eternoendless) [inline js/components/file-input.js]: use `.data` instead of `attr`

```suggestion
      getTextByFilesCount(files, $input.data('multipleFilesText'), $input.data('locale'))
```

**Comment by tomas862** [comment]
@eternoendless done requested changes :)

**Comment by eternoendless** [comment]
Looks good to me!

**Comment by tomas862** [comment]
Changes that are made for this PR in action:

Screenshot

the input on the left has single file upload so it displays the file name that is being uploaded.
the input on the right has multiple files upload functionality and if its more then 1 file it displays the number of files. 

this PR has already compiled assets so the only thing that needs to be done is to checkout to this branch and pull the changes.

go to index.html file to see the changes :+1:

**Comment by matks** [comment]
@tomas862 I'm sorry, I cant see the file numbers being displayed, when I use the custom-file-input. On my localhost it behaves like the regular file input on the left 🤔 

Capture d’écran 2019-03-18 à 15 04 36

**Comment by tomas862** [comment]
on the left yes - because it does not have native html attribute called `multiple`. But for the input on the right, if you select at least 2 files, you will notice the files count :)

**Comment by matks** [comment]
@tomas862 I ran:
```
  530  git clone git@github.com:tomas862/prestashop-ui-kit.git
  531  cd prestashop-ui-kit
  532  git br -a
  533  git checkout features/file-names-in-file-input-type-visibility
```
Then opened index.html, when I use the right input, if I select 1 file I see its name then I select another one it replaces the filename of the 1st

**Comment by matks** [comment]
@tomas862 That was it 😄thank you

**Comment by matks** [comment]
QA said it's 👍 ✅ 🎉 !

**Comment by tomas862** [comment]
nice - so the next step will be to merge this and then someone will create a new version for ui kit and I will be able just to increase the version in prestashop and hit `npm install` again in order to have these changes? :)

**Comment by ntiepresta** [comment]
@tomas862 , @matks 
The adding to display file name on file upload is well resolved with this Pr
Okm

Thanks

**Comment by eternoendless** [comment]
Thank you @tomas862
